Can't build linux-headers on armv6l

**Describe the bug**
`stdenv.__bootPackages.linuxHeaders` (and, therefore, pretty much any package) fails to build on armv6l.

**To Reproduce**
Steps to reproduce the behavior:
1. Get an armv6l machine. (I actually used an aarch64 Raspberry Pi 4B.)
2. `cd path/to/nixpkgs; nix-build . --arg system '"armv6l-linux"' -A stdenv.__bootPackages.linuxHeaders `

```
these derivations will be built:
  /nix/store/6s0a9wyc446qpsn2svia7c9sc91rxbsr-linux-headers-5.9.8.drv
building '/nix/store/6s0a9wyc446qpsn2svia7c9sc91rxbsr-linux-headers-5.9.8.drv' on 'ssh://gaelan@pi4b-00.local'...
unpacking sources
unpacking source archive /nix/store/cai15i0m74ihfajdqfz56xgayc9bf9m3-linux-5.9.8.tar.xz
source root is linux-5.9.8
setting SOURCE_DATE_EPOCH to timestamp 1605039377 of file linux-5.9.8/virt/lib/irqbypass.c
patching sources
applying patch /nix/store/69lyjyca86317hdsc1rwf3ahn7s5kiyn-no-relocs.patch
patching file arch/x86/Makefile
Hunk #1 succeeded at 212 (offset -19 lines).
configuring
no configure script, doing nothing
building
  HOSTCC  scripts/basic/fixdep
bash: line 1:  4020 Segmentation fault      (core dumped) scripts/basic/fixdep scripts/basic/.fixdep.d scripts/basic/fixdep 'gcc -Wp,-MMD,scripts/basic/.fixdep.d -Wall -Wmissing-prototypes -Wstrict-prototypes -O2 -fomit-frame-pointer -std=gnu89         -o scripts/basic/fixdep scripts/basic/fixdep.c   ' > scripts/basic/.fixdep.cmd
make[1]: *** [scripts/Makefile.host:95: scripts/basic/fixdep] Error 139
make[1]: *** Deleting file 'scripts/basic/fixdep'
make: *** [Makefile:539: scripts_basic] Error 2
error: build of '/nix/store/6s0a9wyc446qpsn2svia7c9sc91rxbsr-linux-headers-5.9.8.drv' on 'ssh://gaelan@pi4b-00.local' failed: error: --- Error --- nix-daemon
builder for '/nix/store/6s0a9wyc446qpsn2svia7c9sc91rxbsr-linux-headers-5.9.8.drv' failed with exit code 2; last 10 log lines:
  patching file arch/x86/Makefile
  Hunk #1 succeeded at 212 (offset -19 lines).
  configuring
  no configure script, doing nothing
  building
    HOSTCC  scripts/basic/fixdep
  bash: line 1:  4020 Segmentation fault      (core dumped) scripts/basic/fixdep scripts/basic/.fixdep.d scripts/basic/fixdep 'gcc -Wp,-MMD,scripts/basic/.fixdep.d -Wall -Wmissing-prototypes -Wstrict-prototypes -O2 -fomit-frame-pointer -std=gnu89         -o scripts/basic/fixdep scripts/basic/fixdep.c   ' > scripts/basic/.fixdep.cmd
  make[1]: *** [scripts/Makefile.host:95: scripts/basic/fixdep] Error 139
  make[1]: *** Deleting file 'scripts/basic/fixdep'
  make: *** [Makefile:539: scripts_basic] Error 2
builder for '/nix/store/6s0a9wyc446qpsn2svia7c9sc91rxbsr-linux-headers-5.9.8.drv' failed with exit code 1
error: build of '/nix/store/6s0a9wyc446qpsn2svia7c9sc91rxbsr-linux-headers-5.9.8.drv' failed
```

**Additional context**

The segfault comes from [fixdep](https://github.com/torvalds/linux/blob/master/scripts/basic/fixdep.c), which is invoked by the Linux kernel build system.

Here's a minimal C program that segfaults, when compiled with the `cc` from `nix-shell . --arg system '"armv6l-linux"' -A stdenv.__bootPackages.linuxHeaders`:

```c
#include <stdio.h>
int main(int argc, char** argv)
{
	fprintf(stderr, "Usage: fixdep <depfile> <target> <cmdline>\n");
	return 0;
}

```

A git bisect reveals that the faulty commit is b3640e024f01453b3c4f720135dc6cff529da8ab (#103183).

**Notify maintainers**

@Ericson2314, maintainer of binutils (since that's what seems to be the culprit here)
@expipiplus1, author of #103183

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.
 - system: `"aarch64-linux"`
 - host os: `Linux 4.19.118, NixOS, 21.03.20201220.9cb4d2f (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20201201_5a6ddb3`
 - channels(gaelan): `"nixos-20.09pre225673.8ba41a1e149"`
 - channels(root): `"nixos-20.09pre224185.ed755a65570"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

--- a/pkgs/development/tools/misc/binutils/default.nix
+++ b/pkgs/development/tools/misc/binutils/default.nix
@@ -70,7 +70,7 @@ stdenv.mkDerivation {
        # indeed GHC will refuse to compile with a binutils suffering from it. See
        # this comment for more information:
        # https://gitlab.haskell.org/ghc/ghc/issues/4210#note_78333
-       lib.optional stdenv.targetPlatform.isAarch32 ./R_ARM_COPY.patch;
+       lib.optional (stdenv.targetPlatform.isAarch32 && stdenv.hostPlatform.system != stdenv.targetPlatform.system) ./R_ARM_COPY.patch;
 
   outputs = [ "out" "info" "man" ];
 

